guardian/mobile-n10n
# Add test report to workflows

## What does this change?

This adds test reports to the build workflows. The test reports give us an easier way of visualising what tests have passed and what (if any) have failed. NB: I only added the test reports to projects that actually contain tests.

To ensure the tests generate xml test files I added [test options](https://www.scala-sbt.org/1.x/docs/Testing.html#Test+Framework+Arguments) to the build sbt. In n10n we use specs2 and their [docs](https://etorreborre.github.io/specs2/guide/SPECS2-3.4/org.specs2.guide.JUnitXmlOutput.html) indicate we can generate xml test files using the "junitxml" test option. It was [suggested](https://stackoverflow.com/questions/6934868/configuring-junitxml-output-for-specs2-tests-in-sbt-0-10) to add a "console" test argument to ensure that sbt continues to fail when tests do (I confirmed this locally by forcing tests to fail and validating that the sbt test process fails as expected).

## How to test

After pushing changes we can see test reports generated for the expected projects:

[Football](https://github.com/guardian/mobile-n10n/actions/runs/6010718650/job/16302767233)
[Notification](https://github.com/guardian/mobile-n10n/actions/runs/6010718650/job/16302777182)
[Notification worker lambdas](https://github.com/guardian/mobile-n10n/actions/runs/6010718650/job/16302790335)
[Registration](https://github.com/guardian/mobile-n10n/runs/16302808738)
[Report](https://github.com/guardian/mobile-n10n/actions/runs/6010718650/job/16302773594)
[Schedule lambda](https://github.com/guardian/mobile-n10n/actions/runs/6010718650/job/16302767361)
[SLO monitor](https://github.com/guardian/mobile-n10n/actions/runs/6010909610/job/16303333927)

As a result we can view the test report for each run of a workflow, e.g.:

<img width="1452" alt="Screenshot 2023-08-29 at 13 36 12" src="https://github.com/guardian/mobile-n10n/assets/45561419/27266df3-d95f-4cc4-b4bc-7df51b876b82">
